File: sdk/typescript/src/client/consumer.ts
/**
 * Consumer interface + onMessage pipeline (§1.2 / §1.9 / §A).
 *
 * The real kafkajs adapter is a documented future thin layer that implements
 * `Consumer` and forwards records to `MessagePipeline.onMessage`. This module
 * carries ZERO Kafka dependency — `FakeConsumer` scripts records for tests.
 *
 * Pipeline per record:
 *   1. JSON deserialize (ignore unknown fields, §A).
 *   2. classifySchemaVersion — unknown major (v3+) → reject, do NOT commit.
 *   3. tenant self-check (§1.9) — msg.tenantId === config.tenantId else drop+log.
 *   4. decideBackpressure — in-flight at cap → pause assignment.
 *   5. dispatch to the claim→execute→report flow (handled by lifecycle).
 */

import { classifySchemaVersion, decideBackpressure } from "../decide.ts";

/** A raw dispatch record from the broker (or a fake). */
export interface ConsumerRecord {
  /** raw JSON payload (UTF-8); pipeline deserializes. */
  value: string;
  /** opaque partition/offset bag for commit bookkeeping (optional). */
  meta?: Record<string, unknown>;
}

/** Decoded dispatch message (unknown fields ignored, §A). */
export interface DispatchMessage {
  taskId: string;
  tenantId: string;
  /** routing metadata: the consumer accepts only messages whose workerType it serves. */
  workerType: string;
  schemaVersion?: string | null;
  parameters?: Record<string, unknown>;
  runtimeAttributes?: { traceId?: string } & Record<string, unknown>;
  [k: string]: unknown;
}

/** Kafka assignment control the pipeline can request. */
export interface Assignment {
  pause(): void;
  resume(): void;
  isPaused(): boolean;
}

/** The consumer seam (real kafkajs adapter implements this). */
export interface Consumer {
  /** Begin subscription; deliver each record to the supplied pipeline. */
  start(onMessage: (r: ConsumerRecord) => Promise<void>): Promise<void>;
  /** Interrupt poll (maps to kafkajs `consumer.stop()` / wakeup). */
  wakeup(): Promise<void>;
  pause(): void;
  resume(): void;
  isPaused(): boolean;
}

/** Logger seam (default: console). */
export interface Logger {
  info(msg: string, ctx?: Record<string, unknown>): void;
  warn(msg: string, ctx?: Record<string, unknown>): void;
  error(msg: string, ctx?: Record<string, unknown>): void;
}

export const consoleLogger: Logger = {
  info: (m, c) => console.info(m, c ?? ""),
  warn: (m, c) => console.warn(m, c ?? ""),
  error: (m, c) => console.error(m, c ?? ""),
};

/** Outcome of running one record through the pipeline. */
export type PipelineOutcome =
  | { kind: "accepted"; message: DispatchMessage; committed: boolean }
  | { kind: "rejected-schema"; committed: false }
  | { kind: "dropped-tenant"; committed: false }
  | { kind: "not-for-worker"; committed: false }
  | { kind: "parse-error"; committed: false }
  | { kind: "backpressure"; message: DispatchMessage; committed: false };

export interface MessagePipelineDeps {
  tenantId: string;
  /**
   * workerTypes this worker serves; a message whose `workerType` is not in this
   * set is "not-for-us" and is NOT committed so another worker in the group can
   * reprocess it. Empty / omitted → serve all (no routing filter).
   */
  workerTypes?: readonly string[];
  /** current in-flight task count (live read). */
  inFlight: () => number;
  maxConcurrent: number;
  assignment: Assignment;
  logger?: Logger;
  /** invoked for an accepted, in-tenant, processable message. */
  onAccepted: (msg: DispatchMessage) => Promise<void>;
}

/**
 * Pure-ish pipeline: validates a record and either dispatches it or drops it.
 * "committed" mirrors the offset-commit decision (§A: reject/drop must NOT
 * commit so a fixed deploy can re-read).
 */
export class MessagePipeline {
  #deps: MessagePipelineDeps;
  #logger: Logger;
  #workerTypes: ReadonlySet<string> | undefined;

  constructor(deps: MessagePipelineDeps) {
    this.#deps = deps;
    this.#logger = deps.logger ?? consoleLogger;
    // undefined / empty → no routing filter (serve every workerType).
    this.#workerTypes =
      deps.workerTypes && deps.workerTypes.length > 0
        ? new Set(deps.workerTypes)
        : undefined;
  }

  async onMessage(record: ConsumerRecord): Promise<PipelineOutcome> {
    // 1. deserialize
    let msg: DispatchMessage;
    try {
      msg = JSON.parse(record.value) as DispatchMessage;
    } catch (e) {
      this.#logger.error("dispatch JSON parse failed; not committing", {
        error: String(e),
      });
      return { kind: "parse-error", committed: false };
    }

    // 2. schemaVersion (§A) — unknown major rejected, do not commit
    if (classifySchemaVersion(msg.schemaVersion) === "reject") {
      this.#logger.error("unknown schemaVersion major; rejecting, not committing", {
        schemaVersion: msg.schemaVersion,
        taskId: msg.taskId,
      });
      return { kind: "rejected-schema", committed: false };
    }

    // 3. tenant self-check (§1.9) — last line of defense against ACL drift
    if (msg.tenantId !== this.#deps.tenantId) {
      this.#logger.error("foreign tenant message; dropping, not committing", {
        expected: this.#deps.tenantId,
        actual: msg.tenantId,
        taskId: msg.taskId,
      });
      return { kind: "dropped-tenant", committed: false };
    }

    // 3b. workerType routing — a message not meant for this worker's type must
    // NOT be committed, so another worker in the group reprocesses it.
    if (this.#workerTypes && !this.#workerTypes.has(msg.workerType)) {
      this.#logger.warn("message workerType not served here; not-for-us, not committing", {
        served: [...this.#workerTypes],
        actual: msg.workerType,
        taskId: msg.taskId,
      });
      return { kind: "not-for-worker", committed: false };
    }

    // 4. capacity backpressure
    const bp = decideBackpressure(this.#deps.inFlight(), this.#deps.maxConcurrent);
    if (bp.action === "backpressure") {
      this.#deps.assignment.pause();
      this.#logger.warn("at capacity; pausing assignment (backpressure)", {
        taskId: msg.taskId,
      });
      return { kind: "backpressure", message: msg, committed: false };
    }

    // 5. dispatch to claim→execute→report
    await this.#deps.onAccepted(msg);
    return { kind: "accepted", message: msg, committed: true };
  }
}

/**
 * In-memory consumer for tests: feeds scripted records, records wakeup/pause.
 * No Kafka client.
 */
export class FakeConsumer implements Consumer {
  #records: ConsumerRecord[];
  #paused = false;
  #wokeUp = false;
  #handler?: (r: ConsumerRecord) => Promise<void>;

  constructor(records: ConsumerRecord[] = []) {
    this.#records = records;
  }

  /** Queue more records (e.g. mid-test). */
  feed(...records: ConsumerRecord[]): void {
    this.#records.push(...records);
  }

  async start(onMessage: (r: ConsumerRecord) => Promise<void>): Promise<void> {
    this.#handler = onMessage;
    // drain the scripted queue immediately (cooperative, awaited)
    await this.drain();
  }

  /** Deliver all currently-queued records through the handler. */
  async drain(): Promise<void> {
    if (!this.#handler) return;
    while (this.#records.length > 0 && !this.#wokeUp) {
      const rec = this.#records.shift()!;
      await this.#handler(rec);
    }
  }

  async wakeup(): Promise<void> {
    this.#wokeUp = true;
  }

  get wokeUp(): boolean {
    return this.#wokeUp;
  }

  pause(): void {
    this.#paused = true;
  }

  resume(): void {
    this.#paused = false;
  }

  isPaused(): boolean {
    return this.#paused;
  }
}

/** Default Assignment backed by a Consumer's pause/resume. */
export function assignmentOf(consumer: Consumer): Assignment {
  return {
    pause: () => consumer.pause(),
    resume: () => consumer.resume(),
    isPaused: () => consumer.isPaused(),
  };
}

File: sdk/go/client/consumer.go
package client

import (
	"encoding/json"
	"log"
	"strings"
	"sync"

	"github.com/pinpols/file-batch-system/batch-worker-sdk-go/protocol"
)

// Record is a raw consumed dispatch message (one Kafka record). Value is the
// JSON payload; the real Kafka adapter (a future Consumer impl) fills these.
type Record struct {
	Topic string
	Key   string
	Value []byte
}

// TaskDispatchMessage is the decoded dispatch payload (TaskDispatchMessage
// schema; unknown fields ignored per encoding/json default — wire-protocol §A).
type TaskDispatchMessage struct {
	TaskID            string         `json:"taskId"`
	TenantID          string         `json:"tenantId"`
	SchemaVersion     string         `json:"schemaVersion"`
	WorkerType        string         `json:"workerType"`
	IdempotencyKey    string         `json:"idempotencyKey"`
	RuntimeAttributes map[string]any `json:"runtimeAttributes"`
	Parameters        map[string]any `json:"parameters"`
}

// UnmarshalJSON decodes the dispatch payload, binding the routing key from the
// v2 field `workerType` and falling back to the v1 legacy name `taskType`.
// encoding/json has no alias support (a tag binds one name only), so the v1
// fallback is done explicitly — keeping cross-language parity with Java's
// @JsonAlias("taskType") and Rust's serde alias. Platform v2 only emits
// `workerType`; the fallback is belt-and-suspenders. Guard:
// consumer_test.go TestPipeline_WorkerTypeBinding_*.
func (m *TaskDispatchMessage) UnmarshalJSON(data []byte) error {
	type alias TaskDispatchMessage // shed methods to avoid recursion
	aux := struct {
		*alias
		TaskType string `json:"taskType"`
		// taskId is a BIGINT on the platform, so the orchestrator emits it as a JSON
		// number (e.g. 4515), but the SDK carries it as a string (URL path segment for
		// claim/report). Capture it raw and normalize, tolerating both number and the
		// quoted-string form the contract fixtures historically used. Without this the
		// real dispatch payload fails to decode ("cannot unmarshal number into string").
		TaskID json.RawMessage `json:"taskId"`
	}{alias: (*alias)(m)}
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}
	m.TaskID = strings.Trim(strings.TrimSpace(string(aux.TaskID)), `"`)
	if m.WorkerType == "" && aux.TaskType != "" {
		m.WorkerType = aux.TaskType
	}
	return nil
}

// MessageDisposition is the outcome of the OnMessage pipeline for one record.
type MessageDisposition string

const (
	// DispositionAccepted — message passed all gates; ready to claim+execute.
	DispositionAccepted MessageDisposition = "ACCEPTED"
	// DispositionRejectedSchema — unknown schemaVersion major (§A); offset NOT committed.
	DispositionRejectedSchema MessageDisposition = "REJECTED_SCHEMA"
	// DispositionDroppedForeignTenant — tenant self-check failed (§1.9); dropped.
	DispositionDroppedForeignTenant MessageDisposition = "DROPPED_FOREIGN_TENANT"
	// DispositionDecodeError — payload could not be parsed (poison); offset IS
	// committed (commit-skip) to advance past it, else one corrupt record would
	// head-of-line block the partition forever. fixture 30 / parity §4.5.
	DispositionDecodeError MessageDisposition = "DECODE_ERROR"
	// DispositionBackpressure — at capacity; partition paused, message resumed later (§1.5/§2).
	DispositionBackpressure MessageDisposition = "BACKPRESSURE"
)

// Consumer is the message-source SPI. A real Kafka client (segmentio /
// confluent) plugs in here as a future adapter; the SDK core depends only on
// this interface and never imports a broker library.
type Consumer interface {
	// Poll returns the next batch of records (blocking up to its own timeout) or
	// an empty slice. Returns false when the consumer has been woken/closed.
	Poll() ([]Record, bool)
	// Pause/Resume mirror Kafka assignment.pause()/resume() for backpressure.
	Pause()
	Resume()
	// Commit acknowledges processed offsets.
	Commit()
	// Wakeup interrupts an in-progress Poll (used on graceful stop, §1.6).
	Wakeup()
	// Close releases resources.
	Close()
}

// MessagePipeline runs the per-record gate chain (byo-sdk-guide §1.2/§1.5/§1.9):
// decode -> schemaVersion classify -> tenant self-check -> backpressure.
// It is broker-agnostic: it operates on an already-consumed Record.
type MessagePipeline struct {
	tenantID string
	fsm      *FSM
	logger   *log.Logger
	// inFlight / maxConcurrent drive backpressure; supplied via funcs so the
	// live worker can report real counts.
	inFlight      func() int
	maxConcurrent int
}

// NewMessagePipeline builds a pipeline bound to the worker's tenant + FSM.
func NewMessagePipeline(tenantID string, fsm *FSM, maxConcurrent int, inFlight func() int, logger *log.Logger) *MessagePipeline {
	if logger == nil {
		logger = log.Default()
	}
	if inFlight == nil {
		inFlight = func() int { return 0 }
	}
	return &MessagePipeline{
		tenantID:      tenantID,
		fsm:           fsm,
		logger:        logger,
		inFlight:      inFlight,
		maxConcurrent: maxConcurrent,
	}
}

// OnMessage runs one record through the gate chain. The decoded message is
// returned on DispositionAccepted (and on DispositionBackpressure, since the
// message is valid but deferred). Offset-commit policy is the CALLER's (see
// lifecycle consume loop), keyed on disposition: Accepted + DecodeError commit
// (the latter commit-skip past poison); RejectedSchema / DroppedForeignTenant /
// Backpressure withhold (valid-but-deferred, redeliverable per §1.9 / §A).
func (p *MessagePipeline) OnMessage(rec Record) (TaskDispatchMessage, MessageDisposition) {
	var msg TaskDispatchMessage
	if err := json.Unmarshal(rec.Value, &msg); err != nil {
		p.logger.Printf("ERROR decode dispatch message topic=%s: %v", rec.Topic, err)
		return msg, DispositionDecodeError
	}

	// §A schemaVersion compatibility: unknown major -> reject, do not commit.
	if protocol.ClassifySchemaVersion(msg.SchemaVersion) == "reject" {
		p.logger.Printf("ERROR reject unknown schemaVersion=%q taskId=%s (not committing offset)",
			msg.SchemaVersion, msg.TaskID)
		return msg, DispositionRejectedSchema
	}

	// §1.9 tenant self-check: last line of defense against ACL drift.
	if msg.TenantID != p.tenantID {
		p.logger.Printf("ERROR foreign-tenant message dropped: msgTenant=%q wantTenant=%q taskId=%s",
			msg.TenantID, p.tenantID, msg.TaskID)
		return msg, DispositionDroppedForeignTenant
	}

	// §1.5/§2 capacity backpressure: pause partition when in-flight is full.
	// On the receive path in-flight is at/above max (we just got a new message),
	// so this only ever yields pause; the hysteresis resume path runs as a slot
	// frees (see lifecycle.go), where the decision core gates resume on max/2.
	d := protocol.DecideBackpressure(p.inFlight(), p.maxConcurrent, p.fsm.Paused())
	if d.Action == "backpressure" {
		p.fsm.ApplyKafkaDirective(d.Kafka) // sets paused=true
		p.logger.Printf("INFO backpressure: inFlight=%d max=%d paused partition taskId=%s",
			p.inFlight(), p.maxConcurrent, msg.TaskID)
		return msg, DispositionBackpressure
	}

	return msg, DispositionAccepted
}

// ---------------------------------------------------------------------------
// FakeConsumer — scripted records, no broker (testkit core).
// ---------------------------------------------------------------------------

// FakeConsumer feeds scripted records and records pause/resume/commit/wakeup
// calls. Used by tests in place of a real Kafka client. All methods are
// goroutine-safe because the live worker calls Wakeup()/Close() from Stop while
// the consume loop is still polling — mirroring a real Kafka client's
// thread-safe wakeup contract.
type FakeConsumer struct {
	mu      sync.Mutex
	records [][]Record
	idx     int
	paused  bool
	woken   bool
	Commits int
	Pauses  int
	Resumes int
	Wakeups int
	Closed  bool
}

// NewFakeConsumer scripts successive Poll() batches.
func NewFakeConsumer(batches ...[]Record) *FakeConsumer {
	return &FakeConsumer{records: batches}
}

// Feed appends one more scripted batch.
func (f *FakeConsumer) Feed(batch []Record) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.records = append(f.records, batch)
}

// Poll returns the next scripted batch; ok=false once exhausted or woken.
func (f *FakeConsumer) Poll() ([]Record, bool) {
	f.mu.Lock()
	defer f.mu.Unlock()
	if f.woken {
		return nil, false
	}
	if f.idx >= len(f.records) {
		return nil, false
	}
	batch := f.records[f.idx]
	f.idx++
	return batch, true
}

// Pause records a pause call.
func (f *FakeConsumer) Pause() { f.mu.Lock(); defer f.mu.Unlock(); f.paused = true; f.Pauses++ }

// Resume records a resume call.
func (f *FakeConsumer) Resume() { f.mu.Lock(); defer f.mu.Unlock(); f.paused = false; f.Resumes++ }

// IsPaused reports the fake's pause flag.
func (f *FakeConsumer) IsPaused() bool { f.mu.Lock(); defer f.mu.Unlock(); return f.paused }

// Commit records a commit call.
func (f *FakeConsumer) Commit() { f.mu.Lock(); defer f.mu.Unlock(); f.Commits++ }

// Wakeup records a wakeup and short-circuits further polls.
func (f *FakeConsumer) Wakeup() { f.mu.Lock(); defer f.mu.Unlock(); f.woken = true; f.Wakeups++ }

// Close records a close call.
func (f *FakeConsumer) Close() { f.mu.Lock(); defer f.mu.Unlock(); f.Closed = true }

// counters returns synchronized snapshots for test assertions.
func (f *FakeConsumer) wakeups() int { f.mu.Lock(); defer f.mu.Unlock(); return f.Wakeups }
func (f *FakeConsumer) closed() bool { f.mu.Lock(); defer f.mu.Unlock(); return f.Closed }
func (f *FakeConsumer) commits() int { f.mu.Lock(); defer f.mu.Unlock(); return f.Commits }
